File: validBraces.js
function validBraces(s){
    const stack = []
    const braceMap = {
        ')' : '(',
        ']' : '[',
        '}' : '{'
    };
    const openingBrackets = ['(', '[', '{'];
    const closingBrackets = [')', ']', '}'];

    for(let char of s){
        if(openingBrackets.includes(char)){
            stack.push(char);
        }else if(closingBrackets.includes(char)){
            if(stack.length === 0 || stack.pop() !== braceMap[char]){
                return false;
            }
        }
    }
    return stack.length === 0;

}

console.log(validBraces("()"));
console.log(validBraces("{}{{{}"));
console.log(validBraces("[]"));
console.log(validBraces("()))((()()(("));
console.log(validBraces("[({})]"))

File: valid_braces.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        braceMap = {
            ")" : "(", 
            "}" : "{",
            "]" : "["
        }
        
        for c in s:
            if c in braceMap:
                if stack and stack[-1] == braceMap[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return True if not stack else False
    
#O(n)